### PR TITLE
KL div numerical error fix

### PR DESCRIPTION
**Pull request by v0dro**
Fix for #32520

Related issue #32614 can make the code nicer.

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 67661906:


**None of the build failures appear to be your fault**.

* **2/2** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=6766190618e7529bd9d9711520d958827447f686).

### :snowflake: 2 failures recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4341073) pytorch_linux_xenial_py2_7_9_test (1/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90746258) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90746258)) :snowflake:
<details>
<summary>
<code>Jan 25 09:10:40 AssertionError: tensor(0.3654, grad_fn=<MaxBackward1>) not less than or equal to 1e-05 :</code>
</summary>

```
Jan 25 09:10:40 ---------------------------------------------------------------------- 
Jan 25 09:10:40 Traceback (most recent call last): 
Jan 25 09:10:40   File "test_nn.py", line 8402, in <lambda> 
Jan 25 09:10:40     add(test_name, lambda self, test=test: test(self)) 
Jan 25 09:10:40   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/testing/_internal/common_nn.py", line 3584, in __call__ 
Jan 25 09:10:40     test_case.assertEqual(out, expected_out) 
Jan 25 09:10:40   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/testing/_internal/common_utils.py", line 861, in assertEqual 
Jan 25 09:10:40     assertTensorsEqual(x, y) 
Jan 25 09:10:40   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/testing/_internal/common_utils.py", line 831, in assertTensorsEqual 
Jan 25 09:10:40     self.assertLessEqual(max_err, prec, message) 
Jan 25 09:10:40 AssertionError: tensor(0.3654, grad_fn=<MaxBackward1>) not less than or equal to 1e-05 :  
Jan 25 09:10:40  
Jan 25 09:10:40 ---------------------------------------------------------------------- 
Jan 25 09:10:40 Ran 1497 tests in 465.873s 
Jan 25 09:10:40  
Jan 25 09:10:40 FAILED (failures=6, skipped=709, expected failures=3) 
Jan 25 09:10:40  
Jan 25 09:10:40 Generating XML reports... 
Jan 25 09:10:41 Traceback (most recent call last): 
Jan 25 09:10:41   File "test/run_test.py", line 456, in <module> 
Jan 25 09:10:41     main() 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4341071) pytorch_linux_xenial_py3_6_gcc5_4_test (2/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90746438) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90746438)) :snowflake:
<details>
<summary>
<code>Jan 25 09:10:58 AssertionError: tensor(0.3654, grad_fn=<MaxBackward1>) not less than or equal to 1e-05 :</code>
</summary>

```
Jan 25 09:10:58 ---------------------------------------------------------------------- 
Jan 25 09:10:58 Traceback (most recent call last): 
Jan 25 09:10:58   File "test_nn.py", line 8402, in <lambda> 
Jan 25 09:10:58     add(test_name, lambda self, test=test: test(self)) 
Jan 25 09:10:58   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_nn.py", line 3584, in __call__ 
Jan 25 09:10:58     test_case.assertEqual(out, expected_out) 
Jan 25 09:10:58   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 861, in assertEqual 
Jan 25 09:10:58     assertTensorsEqual(x, y) 
Jan 25 09:10:58   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 831, in assertTensorsEqual 
Jan 25 09:10:58     self.assertLessEqual(max_err, prec, message) 
Jan 25 09:10:58 AssertionError: tensor(0.3654, grad_fn=<MaxBackward1>) not less than or equal to 1e-05 :  
Jan 25 09:10:58  
Jan 25 09:10:58 ---------------------------------------------------------------------- 
Jan 25 09:10:58 Ran 1497 tests in 481.611s 
Jan 25 09:10:58  
Jan 25 09:10:58 FAILED (failures=6, skipped=691, expected failures=2) 
Jan 25 09:10:58  
Jan 25 09:10:58 Generating XML reports... 
Jan 25 09:10:58 Traceback (most recent call last): 
Jan 25 09:10:58   File "test/run_test.py", line 456, in <module> 
Jan 25 09:10:58     main() 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: aten/src/ATen/native/Loss.cpp
Comment (rgommers) [inline aten/src/ATen/native/Loss.cpp]: This is a little ad-hoc and, as your PR description comment indicates, likely better dealt with the same way as gh-31378 (`cosine_similarity`). I suggest to wait till the `cosine_similarity` PR is merged, and then copy that same strategy. Don't try to copy it now, because that PR is still in flux and it'd result in double work.

**Review by rgommers — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/Loss.cpp
Comment (rgommers) [reply aten/src/ATen/native/Loss.cpp]: @v0dro you could also have a chat with @nikitaved about this. I suspect the `cosine_similarity` fix is the right way to go, so if you two can coordinate to push this forward that'd be useful.

Comment (v0dro) [reply aten/src/ATen/native/Loss.cpp]: Alright. I think I'll wait until that PR is merged, understand the methodology and then use that here.

File: aten/src/ATen/native/Loss.cpp
Comment (vadimkantorov) [inline aten/src/ATen/native/Loss.cpp]: Does `zeros` have to be materialized? Can't it be expanded from a scalar tensor value?

**Pull request closed without merging**
